File: scrunch/scripts/scrunch.rb
#!/usr/bin/env ruby
require 'fileutils'
require 'thread'

#Usage : scrunch.rb [--hdfs|--local|--print] job <job args>
# --hdfs: if job ends in ".scala" or ".java" and the file exists, link it against JARFILE (below) and then run it on HOST.
#         else, it is assumed to be a full classname to an item in the JARFILE, which is run on HOST
# --local: run locally according to the rules above
# --print: print the command YOU SHOULD ENTER on the remote node. Useful for screen sessions.

# Configuration in script
##############################################################
SCALA_HOME="/Users/josh/scala/scala-2.9.1.final/lib"
HADOOP_JAR="/Users/josh/cdh/hadoop-0.20.2-cdh3u1/hadoop-core-0.20.2-cdh3u1.jar"

#Get the absolute path of the original (non-symlink) file.
ORIGINAL_FILE=File.symlink?(__FILE__) ? File.readlink(__FILE__) : __FILE__
SCIENCE_ROOT=File.expand_path(File.dirname(ORIGINAL_FILE)+"/../")
JARFILE=SCIENCE_ROOT + "/target/scrunch-1.0-SNAPSHOT-jar-with-dependencies.jar" #what jar has all the depencies for this job
puts JARFILE
TMPDIR="/tmp"
BUILDDIR=TMPDIR+"/script-build"
COMPILE_CMD="java -cp #{SCALA_HOME}/scala-library.jar:#{SCALA_HOME}/scala-compiler.jar -Dscala.home=#{SCALA_HOME} scala.tools.nsc.Main"
##############################################################

if ARGV.size < 1
  $stderr.puts("ERROR: insufficient args.")
  #Make sure to print out up to Configuration above:
  system("head -n 19 #{__FILE__} | tail -n+4")
  exit(1)
end

MODE = case ARGV[0]
  when "--hdfs"
    ARGV.shift
  when "--print"
    ARGV.shift
  else
    #default:
    "--hdfs"
end

JOBFILE=ARGV.shift

def file_type
  JOBFILE =~ /\.(scala|java)$/
  $1
end

def is_file?
  !file_type.nil?
end

PACK_RE = /^package ([^;]+)/
OBJECT_RE = /object\s+([^\s(]+).*/
EXTENSION_RE = /(.*)\.(scala|java)$/

#Get the name of the job from the file.
#the rule is: last class in the file, or the one that matches the filename
def get_job_name(file)
  package = ""
  job = nil
  default = nil
  if file =~ EXTENSION_RE
    default = $1
    File.readlines(file).each { |s|
      if s =~ PACK_RE
        package = $1.chop + "."
      elsif s =~ OBJECT_RE
        unless job and default and (job.downcase == default.downcase)
          #use either the last class, or the one with the same name as the file
          job = $1
        end
      end
    }
    raise "Could not find job name" unless job
    "#{package}#{job}"
  else
    file
  end
end

JARPATH=File.expand_path(JARFILE)
JARBASE=File.basename(JARFILE)
JOBPATH=File.expand_path(JOBFILE)
JOB=get_job_name(JOBFILE)
JOBJAR=JOB+".jar"
JOBJARPATH=TMPDIR+"/"+JOBJAR

def needs_rebuild?
  !File.exists?(JOBJARPATH) || File.stat(JOBJARPATH).mtime < File.stat(JOBPATH).mtime
end

def build_job_jar
  $stderr.puts("compiling " + JOBFILE)
  FileUtils.mkdir_p(BUILDDIR)
  unless system("#{COMPILE_CMD} -classpath #{JARPATH}:#{HADOOP_JAR} -d #{BUILDDIR} #{JOBFILE}")
    FileUtils.rm_rf(BUILDDIR)
    exit(1)
  end

  FileUtils.rm_f(JOBJARPATH)
  system("cp #{JARPATH} #{JOBJARPATH}")
  system("jar uf #{JOBJARPATH} -C #{BUILDDIR} .")
  # FileUtils.rm_rf(BUILDDIR)
end

def hadoop_command
  "hadoop jar #{JOBJARPATH} #{JOB} " + ARGV.join(" ")
end

def jar_mode_command
  "hadoop jar #{JOBJARPATH} #{JOB} " + ARGV.join(" ")
end

if is_file?
  build_job_jar if needs_rebuild?
end

SHELL_COMMAND = case MODE
  when "--hdfs"
    if is_file?
      "#{hadoop_command}"
    else
      "#{jar_mode_command}"
    end
  when "--print"
    if is_file?
      "echo #{hadoop_command}"
    else
      "echo #{jar_mode_command}"
    end
  else
    raise "Unrecognized mode: " + MODE
  end
puts SHELL_COMMAND
system(SHELL_COMMAND)

File: scrunch/scripts/scrunch.py
#!/usr/bin/python

import glob
import os
import re
import shutil
import subprocess
import sys

# Configuration in script
##############################################################
if not "SCALA_HOME" in os.environ:
  sys.stderr.write("Environment variable SCALA_HOME must be set\n")
  sys.exit(1)
SCALA_LIB = os.path.join(os.environ["SCALA_HOME"], "lib")

if not "HADOOP_HOME" in os.environ:
  sys.stderr.write("Environment variable HADOOP_HOME must be set\n")
  sys.exit(1)
HADOOP_HOME = os.environ["HADOOP_HOME"]
HADOOP_JARS = ":".join(glob.glob(os.path.join(HADOOP_HOME, "*.jar")))

#Get the absolute path of the original (non-symlink) file.
if os.path.islink(__file__):
  ORIGINAL_FILE = os.readlink(__file__)
else:
  ORIGINAL_FILE = __file__

SCIENCE_ROOT = os.path.abspath(os.path.dirname(ORIGINAL_FILE)+"/../")
JARFILE = SCIENCE_ROOT + "/target/scrunch-0.2.0-jar-with-dependencies.jar" #what jar has all the depencies for this job
TMPDIR = "/tmp"
BUILDDIR = TMPDIR + "/script-build"
COMPILE_CMD = "java -cp %s/scala-library.jar:%s/scala-compiler.jar -Dscala.home=%s scala.tools.nsc.Main" % (SCALA_LIB, SCALA_LIB, SCALA_LIB)
##############################################################

argv = sys.argv[1:]
if len(argv) < 1:
  sys.stderr.write("ERROR: insufficient args.\n")
  sys.exit(1)

JOBFILE = argv.pop(0)

def file_type():
  m = re.search(r'\.(scala|java)$', JOBFILE)
  if m:
    return m.group(1)
  return None

def is_file():
  return file_type() is not None

PACK_RE = r'package ([^;]+)'
OBJECT_RE = r'object\s+([^\s(]+).*(extends|with)\s+PipelineApp.*'
EXTENSION_RE = r'(.*)\.(scala|java)$'

#Get the name of the job from the file.
#the rule is: last class in the file, or the one that matches the filename
def get_job_name(file):
  package = ""
  job = None
  default = None
  match = re.search(EXTENSION_RE, file)
  if match:
    default = match.group(1)
    for s in open(file, "r"):
      mp = re.search(PACK_RE, s)
      mo = re.search(OBJECT_RE, s)
      if mp:
        package = mp.group(1).trim() + "."
      elif mo:
        if not job or not default or not job.tolower() == default.tolower():
          #use either the last class, or the one with the same name as the file
          job = mo.group(1)
    if not job:
      raise "Could not find job name"
    return "%s%s" % (package, job)
  else:
    return file

JARPATH = os.path.abspath(JARFILE)
if not os.path.exists(JARPATH):
  sys.stderr.write("Scrunch assembly jar not found; run mvn assembly:assembly to construct it.\n")
  sys.exit(1)
  
JARBASE = os.path.basename(JARFILE)
JOBPATH = os.path.abspath(JOBFILE)
JOB = get_job_name(JOBFILE)
JOBJAR = JOB + ".jar"
JOBJARPATH = os.path.join(TMPDIR, JOBJAR)

def needs_rebuild():
  return not os.path.exists(JOBJARPATH) or os.stat(JOBJARPATH).st_mtime < os.stat(JOBPATH).st_mtime

def build_job_jar():
  sys.stderr.write("compiling " + JOBFILE + "\n")
  if os.path.exists(BUILDDIR):
    shutil.rmtree(BUILDDIR)
  os.makedirs(BUILDDIR)
  cmd = "%s -classpath %s:%s -d %s %s" % (COMPILE_CMD, JARPATH, HADOOP_JARS, BUILDDIR, JOBFILE)
  print cmd
  if subprocess.call(cmd, shell=True):
    shutil.rmtree(BUILDDIR)
    sys.exit(1)

  shutil.copy(JARPATH, JOBJARPATH)
  jar_cmd = "jar uf %s -C %s ." % (JOBJARPATH, BUILDDIR)
  subprocess.call(jar_cmd, shell=True)
  shutil.rmtree(BUILDDIR)

def hadoop_command():
  return "%s/bin/hadoop jar %s %s %s" % (HADOOP_HOME, JOBJARPATH, JOB, " ".join(argv))

if is_file() and needs_rebuild():
  build_job_jar()

SHELL_COMMAND = hadoop_command()
print SHELL_COMMAND
os.system(SHELL_COMMAND)
